modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -621,20 +622,53 @@ def map_partitions(
         NumPy array
             An array of partitions
         """
-        preprocessed_map_func = cls.preprocess_func(map_func)
-        return np.array(
-            [
+        if np.prod(partitions.shape) <= 1.5 * CpuCount.get():
+            # block-wise map
+            preprocessed_map_func = cls.preprocess_func(map_func)
+            new_partitions = np.array(
                 [
-                    part.apply(
-                        preprocessed_map_func,
-                        *func_args if func_args is not None else (),
-                        **func_kwargs if func_kwargs is not None else {},
-                    )
-                    for part in row_of_parts
+                    [
+                        part.apply(
+                            preprocessed_map_func,
+                            *func_args if func_args is not None else (),
+                            **func_kwargs if func_kwargs is not None else {},
+                        )
+                        for part in row_of_parts
+                    ]
+                    for row_of_parts in partitions
                 ]
-                for row_of_parts in partitions
-            ]
-        )
+            )
+        else:
+            # axis-wise map
+            # we choose an axis for a combination of partitions
+            # whose size is closer to the number of CPUs
+            if abs(partitions.shape[0] - CpuCount.get()) < abs(
+                partitions.shape[1] - CpuCount.get()
+            ):
+                axis = 1
+            else:
+                axis = 0
+
+            column_splits = CpuCount.get() // partitions.shape[1]
+
+            if axis == 0 and column_splits > 1:
+                # splitting by parts of columnar partitions
+                new_partitions = cls.map_partitions_joined_by_column(
+                    partitions, column_splits, map_func, func_args, func_kwargs
+                )
+            else:
+                # splitting by full axis partitions
+                new_partitions = cls.map_axis_partitions(
```

Comment:
Using `map_axis_partitions` function in `map_partitions` function does not seem obvious and defeats the purpose of `map_partitions` function, which declares that it applies to every partition.

The dataframe level for choosing a suitable strategy seems more appropriate.

Reply:
Since partition mgr rather than core df is designed to play around with partitions, maybe we should just update the docstring?

Reply:
Ok, I added `base_map_partitions` to keep the simplest implementation, but by default we will use new approaches. Are you agree?